src/ImageSharp/Formats/Jpeg/Components/Encoder/ComponentProcessor.cs

```diff
@@ -0,0 +1,233 @@
+            float normalizationValue = -128f;
+
+            int destAreaStride = this.ColorBuffer.Width;
+
+            int yBlockStart = spectralStep * this.component.SamplingFactors.Height;
+
+            Block8x8F workspaceBlock = default;
+
+            // handle subsampling
+            Size subsamplingFactors = this.component.SubSamplingDivisors;
+            if (subsamplingFactors.Width != 1 || subsamplingFactors.Height != 1)
+            {
+                this.PackColorBuffer();
+            }
+
+            int blocksRowsPerStep = this.component.SamplingFactors.Height;
+
+            for (int y = 0; y < blocksRowsPerStep; y++)
+            {
+                int yBuffer = y * this.blockAreaSize.Height;
+                Span<float> colorBufferRow = this.ColorBuffer.DangerousGetRowSpan(yBuffer);
+                Span<Block8x8> blockRow = spectralBuffer.DangerousGetRowSpan(yBlockStart + y);
+                for (int xBlock = 0; xBlock < spectralBuffer.Width; xBlock++)
+                {
+                    // load 8x8 block from 8 pixel strides
+                    int xColorBufferStart = xBlock * 8;
+                    workspaceBlock.ScaledCopyFrom(
+                        ref colorBufferRow[xColorBufferStart],
+                        destAreaStride);
+
+                    // level shift via -128f
+                    workspaceBlock.AddInPlace(normalizationValue);
+
+                    // FDCT
+                    FastFloatingPointDCT.TransformFDCT(ref workspaceBlock);
+
+                    // Quantize and save to spectral blocks
+                    Block8x8F.Quantize(ref workspaceBlock, ref blockRow[xBlock], ref this.quantTable);
+                }
+            }
+        }
+
+        public Span<float> GetColorBufferRowSpan(int row)
+            => this.ColorBuffer.DangerousGetRowSpan(row);
+
+        public void Dispose()
+            => this.ColorBuffer.Dispose();
+
+        private void PackColorBuffer()
+        {
+            Size factors = this.component.SubSamplingDivisors;
+
+            int packedWidth = this.ColorBuffer.Width / factors.Width;
+
+            float averageMultiplier = 1f / (factors.Width * factors.Height);
+            for (int i = 0; i < this.ColorBuffer.Height; i += factors.Height)
+            {
+                Span<float> sourceRow = this.ColorBuffer.DangerousGetRowSpan(i);
+
+                // vertical sum
+                for (int j = 1; j < factors.Height; j++)
+                {
+                    SumVertical(sourceRow, this.ColorBuffer.DangerousGetRowSpan(i + j));
+                }
+
+                // horizontal sum
+                SumHorizontal(sourceRow, factors.Width);
+
+                // calculate average
+                MultiplyToAverage(sourceRow, averageMultiplier);
+
+                // copy to the first 8 slots
+                sourceRow.Slice(0, packedWidth).CopyTo(this.ColorBuffer.DangerousGetRowSpan(i / factors.Height));
```

Comment:
This is one of the reasons 420 subsampling encoding got slower - we abosolutely need to move ready-to-encode color strides to the first 8 indices. This potentially can be resolved at the TPixel -> RGB conversion step - we can rearrange color strides before doing subsampling but it's a very tricky thing to do, I've decided to make this PR as small as possible.